### DISCUSSION TITLE
Latency

### DISCUSSION THREAD

**Discussion by macr023** [probonopd/MiniDexed#385] [General]
I have installed minidexed on a raspberry pi 3b+. Everything works really great with PWM built in sound card + i2c LCD and rotary encoder. I am quite impressed how ever I am experiencing a latency issue compared to hardware synth and even PC software synth. I am runing midi out from my midi keyboard into a korg sq-64 sequencer and I tried both USB from the sequencer directly into the pi and a USB midi dongle. The latency is significant, I don't have specific numbers, how ever I can hear the lag between software synth and minidexed. I even compared against all three,  soft synth (DRC) via usb, hardware synth (reface DX) via midi and minidexed which also connect via same midi splitter.  I tried reducing buffer size to 128 in the default ini file. I played with the settings, I tried different patches. The delay / lag is enough where trying to play in time with an existing beat requires a lot of effort. Its possible but clearly its not natural. Any help is appreciated on how I can reduce latency.

**Comment by probonopd** [comment]
Use a i2s DAC. Day and night.

**Reply by macr023** [reply]
Thanks, I actually have one on order, it should arrive on Thursday. I would of never though that would help with latency/lag

**Reply by probonopd** [reply]
Might also help further to use serial MIDI instead of USB MIDI, although I have not tested due to the lack of being able to make precise latency measurements.

**Reply by macr023** [reply]
If I still have too much latency with the DAC I will give the serial midi approach a shot. Thanks! BTW I tested the usb midi on my laptop running dexed and it feels as instant as my hardware synth itself playing locally.  But I did noticed when I used the headphone jack on the laptop (macbook pro m1) there was pretty bad latency, but out of speakers, it was fine.

**Reply by probonopd** [reply]
Do you have a way for me to measure (not: "feel") latency systematically?

**Reply by macr023** [reply]
I have an oscilloscope,  I can connect one channel to midi in and the other to audio out and set the cursor measurement  between the two.  I can tap into one of the free ports on my midi through box / splitter  for midi signal and just splice a spare 3.5mm cable for the audio side. I will try and make some time today for this, I am curious myself. I will send a screenshot once done.

**Reply by macr023** [reply]
So 

Midi keyboard output into SQ-64 sequencer->
-> sq-64 midi output to 5 port midi through box (splitter)->
->output from midi through box 
                 -> to raspberry pi midi in via USB dongle
                 -> to yamaha reface DX keyboard for hardware synth test

Oscilloscope side
I connected midi pins 4 and 5 to the scope on channel 2 (blue) (On the midi through box / splitter)
I connected 3.5mm audio out to a spliced cable to the scope on channel 1 (yellow)
I also ran a second test where I used audio out form my yamaha DX which plays voice via midi input signal (local is disabled)

Results

- Raspberry PI 3b+ running PWM onboard sound card had an average 14 to 15 ms between midi signal and sound signal
- Yamaha Reface DX with local playback turned off and only plays via midi input had an average of 5 ms latency

oscilloscope screencaps (yeah I know they are photos, I wasn't sure if I can take screencaps and didn't want to spend any more time)

Test with minidexed
[https://photos.app.goo.gl/QJctjnxo4iUcWCRV9](url)

Test with Yamaha reface DX
[https://photos.app.goo.gl/4DH5uFvHmzy2dSxF9](url)


You can see on the bottom right the time between the vertical cursor lines

Am I being a snob complaining over this 9-10ms gap ? I do feel I have to adjust playing and constantly feel if I don't pay attention I will always miss the queue. 

P.S the image links work, but not directly clicking through github, but copying and pasting url will work or should

**Comment by Banana71** [comment]
That is really a long chain of signals. 
I'm really curious what the result would look like with a classic Midi In on the MiniDexed.
A difference of <10ms would be desirable.

The links for the photos don't work for me :-(

**Reply by macr023** [reply]
I have hosted them on my own server


copy and paste url manually in your browser if the link doesn't work
Test with minidexed
[http://www.sorrite.com/midi_test_minidexed.jpg](url)

Test with yamaha reface DX
[http://www.sorrite.com/midi_test_refaceDX.jpg](url)

I connected the yamaha keyboard directly into the raspberry pi and it was lagging a lot compared to local sound being played back, in fact it feels the same. But despite that long chain, the yamaha and the minidexed pi both receive their midi in form the same through box. The reface sounds instant from a keypress. The latency between keyboard midi out to the through box is unknown. I can try and rig something up and measure.

**Reply by macr023** [reply]
So I reran the tests this time this is the order of the equipment


Yamaha Reface DX midi out -> 5 port midi through box (splitter) ->  minidexed + back to reface DX.

It still seems to make no difference since I am measuring midi out from the same point minidexed and the reface DX both are getting their midi in signal. I still noticed the same lag between audio from reface dx VS minidexed. Unfortunately I don't feel like splicing a good midi cable at the moment to pull midi signal directly from keyboard out. How ever this time its almost a direct connection and the lag is still significant. 14.8 to 15ms vs the 4.5 to 5ms using the hardware synth. Tomorrow I will be receiving the i2c DAC, I really hope this will help.

**Comment by Banana71** [comment]
Does it make a difference if you delete the entry "usbspeed=full" in the cmdline.txt file?

**Reply by macr023** [reply]
So I removed usbspeed=full and I noticed some notes not getting triggered, then I realized what was the cause. Turns out using mic in and or line in on your PC is a bad idea. For some reason I plugged my pc speakers directly into minidexed and I was confused. It was perfect, at first I thought it was the usbspeed=full the culprit, but after re-wiring and connecting back through line in, I noticed the lag came back and I was able to connect the dots. It was fun testing this out with the scope.  Thanks for all the help! 

So time to get a mixer.
